### ISSUE TITLE
[1.4] Right clicking an unloaded item to unequip from a dye slot results in an unloaded item with no mod or item name

### ISSUE THREAD

**Issue by blahblahbal** [bluemagic123/tModLoader#2268]
### Version

Latest 1.4-stable

### OS

Windows

### Platform

Steam

### This bug affects

Gameplay as a Player

### Description

I right clicked an unloaded item while equipped in an equipment tab (hooks/pets/etc) dye slot (it became an unloaded item with no mod or item name), and then saved the world, and the game crashed. I have attached the player files in a zip for others to test
[test.zip](https://github.com/tModLoader/tModLoader/files/8459224/test.zip)
.

### Log File

[client.log](https://github.com/tModLoader/tModLoader/files/8459220/client.log)

### Steps to reproduce

1. Load into a world using the test player attached.
2. Switch to equipment tab and right click one of the unloaded items in the dye slots, then put it in your inventory.
3. Attempt to save and quit the world.

### Expected Behavior

The game does not crash

### Actual Behavior

Game crashes

### Reproduction frequency

Every time

### Additional Information

Nothing I can think of.

**Comment by direwolf420** [comment]
Technical breakdown for why this happens:
Assuming a valid unloaded item rests in an equipment dye slot, and it is being right clicked, `ItemSlot.RightClick_FindSpecialActions` is invoked (with context 12), eventually leading to `Main.mouseItem.SetDefaults` being called with the unloaded item type specified. This creates an "empty" `UnloadedItem` which has no data attached, which crashes on saving due to `data` being null:
```cs
public override void SaveData(TagCompound tag) {
	foreach ((string key, object value) in data) {
		tag[key] = value;
	}
}
```
The fix would be to change the SetDefaults to Clone, which makes logical sense as it moves the item from the dye slot to the mouse. (This may also be delegated to Yoraiz0r since Clone is used in various other places with the same context, just not for dyes for some reason)
```cs
//ItemSlot.cs:1472
if (Main.mouseItem.IsAir)
	//Main.mouseItem.SetDefaults(inv[slot].type);
	Main.mouseItem = inv[slot].Clone();
```

There are other places where SetDefaults _may_ be invoked on the unloaded item type too, but atleast in `ItemSlot`, unless someone made `UnloadedItem` researchable or craftable (which is nonsense), there are no such cases.

Still, _should_ `UnloadedItem`s with invalid data on them crash, or not?

**Comment by Solxanich** [comment]
fixed in #2269
